xlators/mount/fuse/src/fuse-bridge.c

```diff
@@ -6065,55 +6329,121 @@ fuse_thread_proc(void *data)
             gf_fuse_mt_iov_base);
 
-        if (!iobuf || !iov_in[0].iov_base) {
+        if (!iobuf_in || !iovec_in[0].iov_base) {
             gf_log(this->name, GF_LOG_ERROR, "Out of memory");
-            if (iobuf)
-                iobuf_unref(iobuf);
-            GF_FREE(iov_in[0].iov_base);
+            if (iobuf_in)
+                iobuf_unref(iobuf_in);
+            GF_FREE(iovec_in[0].iov_base);
             sleep(10);
             continue;
         }
 
-        iov_in[1].iov_base = iobuf->ptr;
+        iovec_in[1].iov_base = iobuf_in->ptr;
+
+        iovec_in[0].iov_len = msg0_size;
+        iovec_in[1].iov_len = psize;
+
+        pthread_mutex_lock(&priv->sq_mutex);
+        {
+            sqe = io_uring_get_sqe(&priv->ring);
+            if (!sqe) {
+                gf_log("this->name", GF_LOG_ERROR,
+                       "Failed to get the submission queue, %s",
+                       strerror(errno));
+
+                iobuf_unref(iobuf_in);
+                GF_FREE(iovec_in[0].iov_base);
+                sleep(10);
+                continue;
+            }
+            io_uring_prep_readv(sqe, priv->fd, iovec_in, 2, /* offset */ 0);
+
+            sqe->flags |= IOSQE_ASYNC;
+            //            sqe->flags |= IOSQE_FIXED_FILE;
+
+            setctx->op = FUSE_READ;
+            setctx->count = 2;
+            setctx->iovec = iovec_in;
+            setctx->iobuf = iobuf_in;
+
+            io_uring_sqe_set_data(sqe, setctx);
+
+        resubmit:
+            ret = io_uring_submit(&priv->ring);
+            if (ret < 0) {
+                gf_log("this->name", GF_LOG_ERROR,
+                       "Failed to submit the request, ret = %d", ret);
+                sleep(10);
+                goto resubmit;
+            }
+        }
+        pthread_mutex_unlock(&priv->sq_mutex);
+
+    wait_cqe:
+        ret = io_uring_wait_cqe(&priv->ring, &cqe);
+        if (ret < 0) {
+            if (ret == -EINTR || ret == -EINVAL) {
+                gf_log("this->name", GF_LOG_WARNING,
+                       "No requests pending, ret = %d", ret);
+                continue;
+            }
+            gf_log("this->name", GF_LOG_ERROR,
+                   "Failed to reap completion queue, ret = %d", ret);
+
+            iobuf_unref(iobuf_in);
+            GF_FREE(iovec_in[0].iov_base);
+            break;
+        }
+
+        fuse_uring_ctx_t *getctx = (fuse_uring_ctx_t *)io_uring_cqe_get_data(
+            cqe);
 
-        iov_in[0].iov_len = msg0_size;
-        iov_in[1].iov_len = psize;
+        iobuf = getctx->iobuf;
+        iov_in = getctx->iovec;
+        int count = getctx->count;
+        int free_count = getctx->free_count;
+        int res = cqe->res;
 
-        res = sys_readv(priv->fd, iov_in, 2);
+        io_uring_cqe_seen(&priv->ring, cqe);
 
-        if (res == -1) {
-            if (errno == ENODEV || errno == EBADF) {
+        if (res < 0) {
+            gf_log("this->name", GF_LOG_WARNING,
```

Comment:
This error comes from the executed operation, not the io_uring queue itself.

Reply:
At line 6396, the check is done for iouring and `res` acts something like errno